internal/app/upstream/multiplexer.go

```diff
@@ -0,0 +1,37 @@
+package multiplexer
+
+import (
+	"context"
+
+	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	"google.golang.org/grpc"
+)
+
+// Multiplexer handles the connections to the upstream management server
+type Multiplexer interface {
+	// QueueRequest starts a stream with the upstream management server
+	// All incoming discovery requests arrive on the request channel
+	// All responses go out through the response channel
+	// All irrecoverable errors go out through the error channel
```

Comment:
Hmm, can we get rid of the error channel for now? I think it creates a disconnect between the request/response that caused the error and the error itself.

What do you think of wrapping `v2.DiscoveryResponse` in a custom struct (or Proto) that defines the discovery response and an optional error? And have a channel of the custom object type instead?

Reply:
I saw https://blog.golang.org/pipelines that using a separate error channel is a pattern. Also it will bee used for irrrecoverable errors. We definitely need a way to bubble up some errors to the caller. The caller could use that to cancel the context and shutdown running goroutines.

Reply:
I can remove the error altogether if you have a strong recommendation for it.

Reply:
+1 to simplify for now and also for using a wrapped struct to carry per-request/response errors.

Reply:
I have removed the error channel for now and wrapping it with the response.